### PR TITLE
WIP global: openaire and fundref harvesting

### PR DESCRIPTION
**Pull request by krzysztof**
- Harvest FundRef and OPENAIRE datasets

### CODE REVIEW

File: invenio_openaire/loaders.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+class RemoteFundRefLoader(BaseFundRefLoader):
+    """Load the FundRef dataset from a remote location."""
+
+    def __init__(self, namespaces=None, cc_resolver=None, source=None):
+        """Initialize the remote loader."""
+        super(RemoteFundRefLoader, self).__init__(
+            namespaces=namespaces, cc_resolver=cc_resolver)
+        self.source = source or \
+            current_app.config['OPENAIRE_FUNDREF_ENDPOINT']
+        resp = urllib2.urlopen(self.source)
```
Comment (lnielsen) [inline invenio_openaire/loaders.py]: You can with advantage use `requests` library instead.

File: invenio_openaire/loaders.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+None of the functions in this module create any objects in invenio-records.
+Instead they handle loading, resolving and finally converting the local XML
+datasets into final JSON that's to be stored using invenio_records.
+
+The OpenAIRE loader comes in two variants: local loader from the pre-fetched
+OpenAIRE dataset using SQLite database and direct remote loader using OAI-PMH
+endpint.
+
+The FundRef dataset loader relies only the local dataset, but it is also
```
Comment (lnielsen) [inline invenio_openaire/loaders.py]: relies only **on** the

File: invenio_openaire/loaders.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+from . import __path__ as current_package
+from .errors import FunderNotFoundError
+
+
+class JSONSchemaResolver(object):
+    """Resolver for '$schema' reference in loaded JSONs."""
+
+    def __init__(self, schemas_host=None, schemas_endpoint=None,
+                 schema_file=None, schema_url=None):
+        """
```
Comment (lnielsen) [inline invenio_openaire/loaders.py]: `"""Create a schema reference.` (no newline)

File: invenio_openaire/ext.py [hunk 4]
```diff
@@ -46,7 +48,6 @@ def init_config(self, app):
             "OPENAIRE_BASE_TEMPLATE",
```
Comment (lnielsen) [inline invenio_openaire/ext.py]: If we're not planning any views then this should just be removed. Same goes for `static`, `templates` and `translations` folder.

**Comment by lnielsen** [comment]
Have you checked the licenses of the fundref and geonames datasets which is included? I believe FundRef is CC0 so that we should be able to include without any problems, but I don't know about geo names.
Also the documentation should state in the License section that these files are not from us.

File: invenio_openaire/config.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+OPENAIRE_OAIPMH_NAMESPACES = {
+    'oai': 'http://www.openarchives.org/OAI/2.0/',
+    'oaf': 'http://namespace.openaire.eu/oaf',
+}
+
+OPENAIRE_SCHEMAS_HOST = 'inveniosoftware.org'
+OPENAIRE_SCHEMAS_ENDPOINT = '/schemas'
+OPENAIRE_SCHEMAS_DEFAULT_FUNDER = 'funder/funder-v1.0.0.json'
+OPENAIRE_SCHEMAS_DEFAULT_GRANT = 'grant/grant-v1.0.0.json'
+OPENAIRE_JSONRESOLVER_GRANTS_HOST = 'inveniogrants.org'
```
Comment (lnielsen) [inline invenio_openaire/config.py]: How is this variable used?

File: invenio_openaire/resolvers/grants.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+from invenio_pidstore.models import PersistentIdentifier
+from invenio_records.api import Record
+from werkzeug.routing import Rule
+
+
+def resolve_grant_endpoint(doi_code, grant_code):
+    """Resolve the OpenAIRE grant."""
+    pid_value = "10.13039/{0}/grants/{1}".format(doi_code, grant_code)
+    pid_type = "recid"
+    pid = PersistentIdentifier.get(pid_type=pid_type, pid_value=pid_value)
```
Comment (lnielsen) [inline invenio_openaire/resolvers/grants.py]: You can use this one: http://pythonhosted.org/invenio-pidstore/usage.html#resolver

``` python
from invenio_pidstore.resolver import Resolver
record = Resolver(pid_type='recid', object_type='rec', getter=Record.get_record).resolve(pid_value)
```

File: invenio_openaire/resolvers/funders.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import jsonresolver
+from invenio_pidstore.models import PersistentIdentifier
+from invenio_records.api import Record
+
+
+@jsonresolver.route('/10.13039/<doi_code>', host='dx.doi.org')
+def resolve_funder(doi_code):
+    """Resolve the JsonRef funder."""
+    pid_value = "10.13039/{}".format(doi_code)
```
Comment (lnielsen) [inline invenio_openaire/resolvers/funders.py:36]: Same as for grants.py (can use the PIDStore resolver)

File: invenio_openaire/jsonschemas/funder/funder-v1.0.0.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{
+  "$schema": "http://json-schema.org/schema#",
```
Comment (lnielsen) [inline invenio_openaire/jsonschemas/funder/funder-v1.0.0.json:2]: We should also provide the Elasticsearch mappings, but that can come afterwards. I'll open an issue.

**Comment by lnielsen** [comment]
Also, remember to add yourself to AUTHORS.rst

**Comment by lnielsen** [comment]
Looks good! Once the Python 3 tests issues and the tiny comments above have been fixed I think it's ready for integration.

File: invenio_openaire/config.py (hunk 5 above)
Comment (krzysztof) [inline invenio_openaire/config.py]: It's used by the JSONResolver for grant resolving here: https://github.com/krzysztof/invenio-openaire/blob/master/invenio_openaire/resolvers/grants.py#L51

It's a more general JSONResolver problem that we don't know a hostname upfront but need to resolve it locally anyway (see JSONSchemas: https://github.com/inveniosoftware/invenio-jsonschemas/blob/master/invenio_jsonschemas/jsonresolver.py#L43).

OpenAIRE resolves two entities using JSONResolver - Funders and Grants.
Funders are straightforward (host = 'dx.doi.org'), while Grants will point to our host, (e.g. 'zenodo.org'), or any other installation. I could have in fact used 'zenodo.org' as default but I wanted something neutral.

File: invenio_openaire/loaders.py (hunk 3 above)
Comment (krzysztof) [inline invenio_openaire/loaders.py]: It seems that no newline invalidates PEP257 - D205

**Pull request merged**